packages/firestore/test/unit/local/indexeddb_schema.test.ts

```diff
@@ -240,6 +254,282 @@ describe('IndexedDbSchema: createOrUpgradeDb', () => {
   });
 });
 
+describe('IndexedDb: canActAsPrimary', () => {
+  if (!IndexedDbPersistence.isAvailable()) {
+    console.warn('No IndexedDB. Skipping canActAsPrimary() tests.');
+    return;
+  }
+
+  async function clearOwner(): Promise<void> {
+    const simpleDb = await SimpleDb.openOrCreate(
+      INDEXEDDB_TEST_DATABASE,
+      SCHEMA_VERSION,
+      createOrUpgradeDb
+    );
+    await simpleDb.runTransaction('readwrite', [DbOwner.store], txn => {
+      const ownerStore = txn.store<DbOwnerKey, DbOwner>(DbOwner.store);
+      return ownerStore.delete('owner');
+    });
+    simpleDb.close();
+  }
+
+  beforeEach(() => {
+    return SimpleDb.delete(INDEXEDDB_TEST_DATABASE);
+  });
+
+  after(() => SimpleDb.delete(INDEXEDDB_TEST_DATABASE));
+
+  interface PrimaryStateTestCases {
+    thisClient: { networkEnabled: boolean; visibilityState: VisibilityState };
+    thatClient: { networkEnabled: boolean; visibilityState: VisibilityState };
+    expectedPrimaryState: boolean;
+  }
+
+  const testCases: PrimaryStateTestCases[] = [
+    {
+      thatClient: {
+        networkEnabled: false,
+        visibilityState: 'hidden'
+      },
+      thisClient: {
+        networkEnabled: false,
+        visibilityState: 'hidden'
+      },
+      expectedPrimaryState: true
+    },
```

Comment:
Hrm...   Not sure how I feel about manually listing the 16 different possible cases and the expected result (using 11 lines of "code" for each one)...  I wonder:

1) Do we really need to test this that exhaustively?
2) If yes, could we use a smaller template?  e.g.:

```
// Test Cases
// [our-network, our-visibility, their-network, their-visibility, expectedPrimaryState]
[
  [false, 'hidden', false, 'hidden' true],
  ...
];
```

I guess I don't care that much, but as-is this is difficult to read (I'm not going to compare all of these to make sure you got all the cases) and having 300 lines of code to test ~50 lines seems questionable.